blocks/components/editable/index.js

```diff
@@ -306,14 +305,35 @@ export default class Editable extends wp.element.Component {
 		return !! this.state.formats[ format ];
 	}
 
-	toggleFormat( format ) {
-		this.editor.focus();
-
-		if ( this.isFormatActive( format ) ) {
-			this.editor.formatter.remove( format );
-		} else {
-			this.editor.formatter.apply( format );
+	changeFormats( formats ) {
+		if ( this.state.bookmark ) {
+			this.editor.selection.moveToBookmark( this.state.bookmark );
 		}
+
+		forEach( formats, ( formatValue, format ) => {
+			if ( format === 'link' ) {
+				if ( formatValue !== undefined ) {
+					const anchor = this.editor.dom.getParent( this.editor.selection.getNode(), 'a' );
+					if ( ! anchor ) {
+						this.editor.formatter.remove( 'link' );
+					}
+					this.editor.formatter.apply( 'link', { href: formatValue.value }, anchor );
+				} else {
+					this.editor.execCommand( 'Unlink' );
+				}
+			} else {
+				const isActive = this.isFormatActive( format );
+				if ( isActive && ! formatValue ) {
+					this.editor.formatter.remove( format );
+				} else if ( ! isActive && formatValue ) {
+					this.editor.formatter.apply( format );
+				}
+			}
+		} );
+
+		this.setState( {
+			formats: merge( this.state.formats, formats )
```

Comment:
You're mutating `this.state` directly here. Consider passing empty object `{}` as first argument instead.

>Never mutate `this.state` directly, as calling `setState()` afterwards may replace the mutation you made. Treat `this.state` as if it were immutable.

https://facebook.github.io/react/docs/react-component.html#state

>**Note**: This method mutates `object`.

https://lodash.com/docs/4.17.4#merge